crates/wasm-encoder/src/core/dump.rs

```diff
@@ -0,0 +1,228 @@
+use std::borrow::Cow;
+
+use crate::{CustomSection, Encode, Section};
+
+/// The "core" custom section for coredumps, as described in the
+/// [tool-conventions
+/// repository](https://github.com/WebAssembly/tool-conventions/blob/main/Coredump.md)
+///
+/// # Example
+///
+/// ```
+/// use wasm_encoder::{CoreDumpSection, Module};
+/// let core = CoreDumpSection::new("MyModule.wasm");
+/// let mut module = Module::new();
+/// module.section(&core);
+/// ```
+#[derive(Clone, Debug, Default)]
+pub struct CoreDumpSection {
+    name: String,
+}
+
+impl CoreDumpSection {
+    /// Create a new core dump section encoder
+    pub fn new(name: impl Into<String>) -> Self {
+        let name = name.into();
+        CoreDumpSection { name }
+    }
+}
+
+impl Encode for CoreDumpSection {
+    fn encode(&self, sink: &mut Vec<u8>) {
+        let mut data = vec![0];
+        self.name.encode(&mut data);
+        CustomSection {
+            name: "core".into(),
+            data: Cow::Owned(data),
+        }
+        .encode(sink);
+    }
+}
+
+impl Section for CoreDumpSection {
+    fn id(&self) -> u8 {
+        crate::core::SectionId::Custom as u8
+    }
+}
+
+/// A "corestack" custom section as described in the [tool-conventions
+/// repository](https://github.com/WebAssembly/tool-conventions/blob/main/Coredump.md)
+///
+/// Example
+///
+/// ```
+/// use wasm_encoder::{CoreDumpStackSection, Module, CoreDumpValue};
+/// let mut thread = CoreDumpStackSection::new("main");
+/// thread.frame(42, 1000, vec![CoreDumpValue::I32(1)], vec![CoreDumpValue::I32(2)]);
+/// let mut module = Module::new();
+/// module.section(&thread);
+/// ```
+#[derive(Clone, Debug, Default)]
+pub struct CoreDumpStackSection {
+    frame_bytes: Vec<u8>,
+    count: u32,
+    name: String,
+}
+
+impl CoreDumpStackSection {
+    /// Create a new core dump stack section encoder
+    pub fn new(name: impl Into<String>) -> Self {
+        let name = name.into();
+        CoreDumpStackSection {
+            frame_bytes: Vec::new(),
+            count: 0,
+            name,
+        }
+    }
+
+    /// Encode a stack frame and add it to the encoded bytes of the stack section
```

Comment:
Nitpick: I think the docs are a little more concise and clear when worded like this:

```suggestion
    /// Add a stack frame to this coredump stack section.
```